zebra-network/src/peer_set/initialize.rs

```diff
@@ -189,30 +200,54 @@ where
 }
 
 /// Use the provided `handshaker` to connect to `initial_peers`, then send
-/// the results over `tx`.
-#[instrument(skip(initial_peers, outbound_connector, tx))]
+/// the results over `peer_set_sender`.
+///
+/// Updates the `AddressBook` using `timestamp_collector`.
+#[instrument(skip(
+    initial_peers,
+    outbound_connector,
+    timestamp_collector,
+    peer_set_sender
+))]
 async fn add_initial_peers<S>(
     initial_peers: std::collections::HashSet<SocketAddr>,
     outbound_connector: S,
-    mut tx: mpsc::Sender<PeerChange>,
+    mut timestamp_collector: mpsc::Sender<MetaAddrChange>,
+    mut peer_set_sender: mpsc::Sender<PeerChange>,
 ) -> Result<(), BoxError>
 where
     S: Service<SocketAddr, Response = Change<SocketAddr, peer::Client>, Error = BoxError> + Clone,
     S::Future: Send + 'static,
 {
     info!(?initial_peers, "connecting to initial peer set");
-    // ## Correctness:
+
+    // Add the DNS seeder peers to the address book, in the `AttemptPending` state
+    for addr in initial_peers.iter() {
+        // Treat DNS seeder peers like gossiped peers, but fill in defaults as needed
+        // TODO: make unique State or Change variants for DNS seeder peers?
+        let meta_addr =
+            MetaAddr::new_gossiped_meta_addr(&addr, &PeerServices::NODE_NETWORK, &Utc::now());
+        let _ = timestamp_collector
+            .send(MetaAddr::new_gossiped_change(&meta_addr))
+            .await;
+        let _ = timestamp_collector
+            .send(MetaAddr::update_attempt(&addr))
+            .await;
+    }
+
+    // # Correctness
     //
-    // Each `CallAll` can hold one `Buffer` or `Batch` reservation for
-    // an indefinite period. We can use `CallAllUnordered` without filling
+    // Each `FuturesUnordered` can hold one `Buffer` or `Batch` reservation for
+    // an indefinite period. We can use `FuturesUnordered` without filling
     // the underlying `Inbound` buffer, because we immediately drive this
-    // single `CallAll` to completion, and handshakes have a short timeout.
+    // single `FuturesUnordered` to completion, and handshakes have a short
+    // timeout.
     let mut handshakes: FuturesUnordered<_> = initial_peers
-        .into_iter()
+        .iter()
         .map(|addr| {
             outbound_connector
                 .clone()
-                .oneshot(addr)
+                .oneshot(*addr)
                 .map_err(move |e| (addr, e))
```

Comment:
TODO: spawn a separate task for each of these handshakes.

This is like #1950, but in the initial handshakes.

Reply:
Can we just spawn a dial task here?
We don't want to put errors back in the demand queue, but otherwise we want basically the same behaviour.

Reply:
a194f982ac125c9d13ca2ed92e368c20da33bd17